dttools/src/jx_function.c

```diff
@@ -551,262 +550,323 @@ static struct jx *expand_template(struct jx *template, struct jx *ctx, struct jx
-					goto FAILURE;
-				}
-				result = jx_parse_stream(stream);
-				pclose(stream);
-			} else {
-				FILE *stream = fopen(path, "r");
-				if(!stream) {
-					err = string_format("error reading %s: %s\n", path,strerror(errno));
-					goto FAILURE;
-				}
-				result = jx_parse_stream(stream);
-				fclose(stream);
-			}
+	if (!jx_istype(val, JX_STRING)) {
+		result = make_error(func, args, " string argument required");
+		goto FAILURE;
+	}
 
-			if(!result) {
-				err = string_format("error parsing JSON in %s",path);
-				goto FAILURE;
-			}
-			break;
-		default:
-			err = string_format("error fetch() expects a string argument");
+	const char *path = val->u.string_value;
+	if(string_match_regex(path, "http(s)?://")) {
+		//Arbitrary 30 second timeout to perform curl
+		char *cmd = string_format("curl -m 30 -s %s", path);
+		FILE *stream = popen(cmd,"r");
+		free(cmd);
+		if(!stream) {
+			result = make_error(func, args, "error fetching %s: %s",
+				path, strerror(errno));
 			goto FAILURE;
+		}
+		result = jx_parse_stream(stream);
+		pclose(stream);
+	} else {
+		FILE *stream = fopen(path, "r");
+		if(!stream) {
+			result = make_error(func, args, "error reading %s: %s\n",
+				path, strerror(errno));
+			goto FAILURE;
+		}
+		result = jx_parse_stream(stream);
+		fclose(stream);
 	}
+
+	if(!result) {
+		result = make_error(func, args, "error parsing JSON in %s", path);
+		goto FAILURE;
+	}
+
+FAILURE:
 	jx_delete(args);
 	jx_delete(val);
 	return result;
-	
-	FAILURE:
-	jx_delete(args);
-	jx_delete(val);
-	FAIL(funcname, orig_args, err);
 }
 
-struct jx *jx_function_select(struct jx *orig_args, struct jx *ctx) {
-	assert(orig_args);
+struct jx *jx_function_select(struct jx *args, struct jx *ctx) {
+	assert(args);
+	if (jx_istype(args, JX_ERROR)) return args;
+	assert(jx_istype(args, JX_ARRAY));
 	assert(jx_istype(ctx, JX_OBJECT));
-	const char *funcname = "select";
-	const char *err = NULL;
+	const char *func = "select";
 
-	//Get args and parse stringified query
-	//Stringify performed during initial eval
-	struct jx *args = jx_copy(orig_args);
-	struct jx *val = jx_parse_string(jx_array_shift(args)->u.string_value);
+	struct jx *result = NULL;
+	struct jx *new_ctx = NULL;
+	struct jx *j = NULL;
+
+	struct jx *val = jx_array_shift(args);
 	struct jx *objlist = jx_array_shift(args);
-	assert(jx_istype(objlist, JX_ARRAY));
-	struct jx *result = jx_array(0);
+	if (jx_array_length(args) != 0) {
+		result = make_error(func, args, "2 arguments required");
+		goto FAILURE;
+	}
+	if (!jx_istype(objlist, JX_ARRAY)) {
+		result = make_error(func, args, "list of objects required");
+		goto FAILURE;
+	}
+
+	result = jx_array(0);
 
 	struct jx *item;
 	for(void *i = NULL; (item = jx_iterate_array(objlist, &i));) {
-		struct jx *j = jx_eval(val, jx_merge(ctx, item, NULL));
-		assert(jx_istype(j, JX_BOOLEAN));
-		if(!j) {
-			err = "error evaluating select expression";
+		if (!jx_istype(item, JX_OBJECT)) {
```

Comment:
Forgot those functions *do* eval the second arg. Should be fixed

Reply:
One step closer:

```
Enter context expression (or {} for an empty context):
{};
Now enter expressions:
select(like(".*work-queue.*", type), [fetch("/disk/d11/tlq/0483083f-dc09-4426-a0d8-3374b02e7da3.json")]);
expression: select(like(".*work-queue.*",type),[fetch("/disk/d11/tlq/0483083f-dc09-4426-a0d8-3374b02e7da3.json")])
value:   
```

We should expect this to return `[ ]`.

```
select("type" == "test", 
[fetch("/disk/d11/tlq/0483083f-dc09-4426-a0d8-3374b02e7da3.json")]);
expression: select("type"=="test",[fetch("/disk/d11/tlq/0483083f-dc09-4426-a0d8-3374b02e7da3.json")])
Segmentation fault
```

Trying to quote a key (`"type"`) used to return a symbol lookup error I believe.

Reply:
Hmmm.... using `"type"=="test"` should give an empty result (since the two strings are never equal), but definitely not a segfault. Mind posting a backtrace?